### ISSUE TITLE
How to best represent future in RAC3

### ISSUE THREAD

**Issue by tonyxiao** [ReactiveCocoa/ReactiveCocoa#2220]
What's the best way to semantically represent a `future` in RAC3? `Signal` doesn't cut it because it has no concept of current value and if the underlying work has completed it won't send the value again to new observers. `Property` has to have an initial value and isn't suited either. From a technical point of view only `SignalProducer.buffer` is capable of representing the idea of a future but it's a real pain to determine whether the future has been fulfilled and semantically it also feels wrong to call `start` on the `SignalProducer` to get values because  the underlying work has already been started and calling start won't start the work again. 

Personally I've been adding a new data type called `Future` that's very similar to `PropertyOf` and has a `value?`, `producer` and a number of other convenience methods. However I'm interested in what everyone else does and considers to be best practice.

**Comment by nghialv** [comment]
:+1:

**Comment by russbishop** [comment]
I'm trying to figure out the very same thing. ReactiveCocoa seems well suited to the "hot" signal case where signals represent a stream of externally-induced events over time. It also handles a few of the "cold" signal cases like a network request.

What doesn't make as much sense to me is something like an asynchronous database query or creating an async wrapper around `NSFileManager`. It feels wrong to return a producer in those cases because it's asking the caller to then turn around and call `start`. In a great many cases I want any chained observers to simply observe the last value received (often the only value) or observe the error, whether they are connected later or not. I'm not sold that separating into `Signal` and `SignalProducer` is actually a net win.

I started implementing a `Future<T>` using RAC underneath, then realized I was just starting to re-implement PromiseKit and bailed. Either I'm not buying into RAC's philosophy correctly or the design is too obtuse and complicated. I fully admit it could be my failure to grasp the new design. 

If anyone has a handy example of how you'd wrap something like `NSData.writeToURL` so an asynchronous observer can know when the write completes or errors I would appreciate it.

**Comment by nomothetis** [comment]
@tonyxiao I'm not sure why it "feels wrong" to call `start`, but it's definitely incorrect to say that " the underlying work has already been started and calling start won't start the work again." It actually *will*; that's the whole point of the producer.

@russbishop Unless I'm misunderstanding, it seems like `SignalProducer` does make sense in the `NSFileManager` case. A `SignalProducer` is a representation of a computation, not an actual computation, so calling `start` is expected. I'm not exactly following what you mean by chained observers observing things "whether they are connected or not", but there are definitely functions that allow you to wait until the end and take the last value (`last`, `taksLast`), or to only observe the error (call `start` with only an `error:` parameter). Do you have a more specific use case?

I'm not entirely sure, incidentally, that `Future` is in the same category of abstractions as `Signal` and `SignalProducer`, so you may in fact need to create a new one for that. That said `Signal` and `SignalProducer` may simply provide a *different* way of achieving the goals of `Future`?

**Comment by neilpa** [comment]
Agree with @nomothetis on all his points and would also recommend looking at the [`CHANGELOG`](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/master/CHANGELOG.md). Specifically the [replacements](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/master/CHANGELOG.md#replacements) section but imagine `Future` wherever it says `RACSignal` because it has basically the same set of problems.

**Comment by russbishop** [comment]
@neilpa @nomothetis So in the `NSFileManager` case how does it make sense to `start` a `SignalProducer` setup to write the same data to the same file multiple times? That's what I'm struggling with... I want the function to control that and only allow the caller to observe the results. Maybe that's not really a problem in practice? 

I think you both are probably correct... I'm thinking in terms of promises and RAC has a different philosophy.

**Comment by nomothetis** [comment]
Ah, I see what you're trying to do. Without taking the time to read up on the `NSFileManager` API, here's the general approach. First, have a function that creates a signal producer for your data/file combination:

```swift
func producerToSaveData(data:NSData, toFile file:NSURL) -> SignalProducer<Void, NSError> {
    return SignalProducer { sink, disposable in
        // assume async API roughly like this
        NSFileManager.saveData(data, toURL:file) { error in
            if error == nil {
                sendCompleted(sink)
            } else {
                sendError(error, sink)
            }
        }
    }
}
```

You then use the function in your remaining code like this:

```swift
let data = …//data to save
let path = …//path
let disposable = producerToSaveData(data, toFile:path).start(error: { err in
    NSLog("Failed to save due to error: \(err)")
})
```

(written in this text field, so please excuse typos, compile errors, etc.)

Hope this helps.

**Comment by tonyxiao** [comment]
What's not easy in the above file manager example. 

1) Given a `SignalProducer<Void, Error>`, how can I easily tell whether the filemanager operation has completed or still in progress? I would need to do something like

```
var completed = false
producer.start(completed: {
    completed = true
})
if completed {
    // do something
} else {
    // do something else
}
```

2) Suppose now that I start the signal producer, do something with it, and want to return the the future result of this computation to potentially interested parties, regardless if anyone is interested. I can't return the producer, because I don't want the underlying work to be started again, what matters is the result of my transformation. I can't return a Signal, because if the work is already done by the time someone observes it they will get `interrupt` event instead of useful value. The best thing I can do in current RAC lingo is to return a buffered signal producer. But then we are back in the dilemma of "how do I tell if this is gonna start the underlying network request?"

Basically I think it comes down to

Given that an expensive operation which is going to deliver exactly one value has already be started, but we do not yet know whether or not it has completed, how do we represent this in RAC?

**Comment by nomothetis** [comment]
What you're asking for is basically to introduce side-effects on an already started signal. The way to do that is to use the `on` method like so:

```swift
let producer = …//create SignalProducer

producer.start(…)

// pass the producer around to anyone interested. When they're interested, they call:
producer.on(start: { … },
            next: { … },
            error: { … },
            interrupted: { … }
            completed: { … })
```

Hope this helps.

**Comment by nomothetis** [comment]
I agree that those things you want to do are difficult. I'm back to thinking that the *specific* objective of your code may be achievable with RAC, but that this particular way of going about it is not.

**Comment by mnmaraes** [comment]
A solution may be `PropertyOf<FutureValue<ValueType>>`
where:
```
 enum FutureValue<T> {
    case NoValue
    indirect case Value(T) // Or: case Value(Box<T>)
}
```
That seems to satisfy everything you asked for.

**Comment by neilpa** [comment]
@mnmaraes `FutureValue` is just `Optional` so you could simplify that to `PropertyOf<ValueType?>`.

I'll add that I've almost written `Deferred<T>`/`Fault<T>` several times. However, since that's essentially `PropertyOf<T?>` I just ended up using that.

**Comment by mnmaraes** [comment]
@neilpa Very True. `PropertyOf<T?>` has always sufficed for me too.

There might advantages to using a different name for the same abstraction, though. Especially when trying to be didactic.

For example:
`value == .NoValue` is checking for completion 
whereas 
`value == nil` may be checking for completion or checking if the fulfilled value of the future is `nil`.

**Comment by neilpa** [comment]
In that case you would have `T??` but that can get confusing.
